computedfields/resolver.py

```diff
@@ -0,0 +1,661 @@
+        return f
+    if not wrapped_save_methods.get(f):
+        wrapped_save_methods[f] = wrap
+    return wrapped_save_methods[f]
+
+def undecorate_save(f):
+    for orig, wrapped in wrapped_save_methods.items():
+        if wrapped == f:
+            return orig
+    return f
+
+
+# replacement for isinstance(model, ComputedFieldsModelType)
+# TODO: move to resolver
+def has_computedfields(model):
+    return model in Resolver._computed_models
+
+
+# TODO: make this instance based with DEFAULT_RESOLVER and custom resolvers?
+class ResolverBase:
+    """
+    Holds the needed data for graph calculations and dependency resolving.
+
+    After startup the method ``_resolve_dependencies`` gets called by
+    ``app.ready`` to build the dependency resolving map.
+    To avoid the expensive calculations in production mode the map
+    can be pickled into a map file by setting ``COMPUTEDFIELDS_MAP``
+    in settings.py to a writable file path and calling the management
+    command ``createmap``.
+
+    .. NOTE::
+
+        The map file will not be updated automatically and therefore
+        must be recreated by calling the management command
+        ``createmap`` after model changes.
+    """
+    _lock = RLock()
+
+
+    def __init__(self):
+        self._graph = None
+        self._computed_models = {}
+        self._map = {}
+        self._fk_map = {}
+        self._local_mro = {}
+        self._map_loaded = False
+        self._batchsize = getattr(settings, 'COMPUTEDFIELDS_BATCHSIZE', 100)
+
+    def cf_mro(self, model, update_fields=None):
+        """
+        Return mro for local computed field methods for a given set of ``update_fields``.
+        This method returns computed fields as self dependent to simplify field calculation in ``save``.
+        """
+        # TODO: investigate - memoization of update_fields result? (runs ~4 times faster)
+        entry = self._local_mro[model]  # raise here, if model is not a CMFT - FIXME: can happen, if model has no cf
+        if update_fields is None:
+            return entry['base']
+        update_fields = frozenset(update_fields)
+        base = entry['base']
+        fields = entry['fields']
+        mro = 0
+        for f in update_fields:
+            mro |= fields.get(f, 0)
+        return [name for pos, name in enumerate(base) if mro & (1 << pos)]
+    
+    def extract_computed_models(self):
+        # pull _computed_fields from collector
+        computed_models = {}
+        for model, computedfields in ACTIVE_COLLECTOR.models_with_computedfields:
+            if not computedfields:
+                continue
+            computed_models[model] = {}
+            _computed_fields = {}       # FIXME: remove from codebase
+            for field in computedfields:
+                computed_models[model][field.name] = field._computed['depends']
+                _computed_fields[field.name] = field
+            model.save = decorate_save(model.save)
```

Comment:
Why did you choose this solution vs. registering a `post_save` signal on all models that have computed fields ?

Reply:
Thats coupled to an issue, that I never phrased into an issue report. In the old code `save` overwrites will still hold dirty cf values, which basically makes them unusable for custom handling in your own `save` method. The decorator pattern solves this.

Not implemented yet: If the decorator finds a custom save method it shall do the final cf calcs again afterwards. Not sure yet about that one, maybe it is easier to skip that (perfwise) and place a note in the docs, how to sync cf values after own field modifications in save.

Overall this is the only way I could find to make cf values usable in save and ensure that the DB still gets synced values.

Note that `post_save` is to late to do that (values already written to DB), unless the record gets saved a second time.